sdks/python/src/lilypad/_internal/otel/stream_wrappers.py

```diff
@@ -0,0 +1,358 @@
+            metadata, choice_deltas = self.process_chunk(chunk)
+            self._update(metadata, choice_deltas)
+            return chunk
+        except StopIteration:
+            self._cleanup()
+            raise
+        except Exception as error:  # pragma: no cover
+            self.span.set_status(Status(StatusCode.ERROR, str(error)))
+            self.span.set_attribute(
+                error_attributes.ERROR_TYPE, type(error).__qualname__
+            )
+            self._cleanup()
+            raise
+
+
+class AsyncStreamWrapper(BaseStreamWrapper[ChunkT], Generic[ChunkT]):
+    """Asynchronous stream wrapper with async context manager support."""
+
+    stream: AsyncStream[ChunkT]
+
+    def __init__(
+        self,
+        span: Span,
+        stream: AsyncStream[ChunkT],
+        process_chunk: ProcessChunk[ChunkT],
+    ) -> None:
+        """Initialize the stream wrapper with telemetry components."""
+        self.span = span
+        self.stream = stream
+        self.process_chunk = process_chunk
+        self.response_attributes = GenAIResponseAttributes()
+        self.choice_buffers: list[ChoiceBuffer] = []
+        self._span_started = False
+        self._setup()
+
+    async def __aenter__(self) -> "AsyncStreamWrapper[ChunkT]":  # pragma: no cover
+        """Enter the async context manager."""
+        self._setup()
+        return self
+
+    async def __aexit__(
+        self,
+        exc_type: type[BaseException] | None,
+        exc_val: BaseException | None,
+        exc_tb: TracebackType | None,
+    ) -> bool:  # pragma: no cover
+        """Exit the async context manager and handle exceptions."""
+        try:
+            if exc_type is not None:
+                self.span.set_status(Status(StatusCode.ERROR, str(exc_val)))
+                self.span.set_attribute(
+                    error_attributes.ERROR_TYPE, exc_type.__qualname__
+                )
+        finally:
+            self._cleanup()
+        return False
+
+    async def close(self) -> None:  # pragma: no cover
+        """Close the async stream and perform cleanup."""
+        await self.stream.close()
+        self._cleanup()
+
+    def __aiter__(self) -> "AsyncStreamWrapper[ChunkT]":
+        """Returns self as an async iterator."""
+        return self
+
+    async def __anext__(self) -> ChunkT:
+        """Get and process the next chunk from the async stream."""
+        try:
+            chunk = await self.stream.__anext__()
+            metadata, choice_deltas = self.process_chunk(chunk)
+            self._update(metadata, choice_deltas)
+            return chunk
+        except StopAsyncIteration:
+            self._cleanup()
+            raise
+        except Exception as error:  # pragma: no cover
+            self.span.set_status(Status(StatusCode.ERROR, str(error)))
+            self.span.set_attribute(
+                error_attributes.ERROR_TYPE, type(error).__qualname__
+            )
+            self._cleanup()
```

Comment:
I'm noting this logic is duplicated 2x per class, wonder if this could also become a helper method on the base class